pkg/localmetrics/localmetrics.go

```diff
@@ -16,51 +16,167 @@ package localmetrics
+		c.Collect(ch)
+	}
+}
 
-	req, err := http.NewRequest("GET", "https://api.deadmanssnitch.com/v1/snitches", nil)
-	if err != nil {
-		MetricDeadMansSnitchHeartbeat.With(prometheus.Labels{"name": "deadmanssnitch-operator"}).Set(float64(0))
+func NewMetricsCollector() *MetricsCollector {
+	return &MetricsCollector{
+		ReconcileDuration: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:        "dms_operator_reconcile_duration_seconds",
+			Help:        "The duration it takes to reconcile a ClusterDeployment",
+			ConstLabels: map[string]string{"name": operatorName},
+		}),
+		// apiCallDuration times API requests. Histogram also gives us a _count metric for free but also includes errors.
+		apiCallDuration: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Name:        "dms_operator_api_request_duration_seconds",
+			Help:        "Distribution of the number of seconds an API request takes",
+			ConstLabels: prometheus.Labels{"name": operatorName},
+			// We really don't care about quantiles, but omitting Buckets results in defaults.
+			// This minimizes the number of unused data points we store.
+			Buckets: []float64{1},
+		}, []string{"controller", "method", "resource", "status"}),
+		snitchCallErrors: prometheus.NewCounter(prometheus.CounterOpts{
+			Name:        "dms_operator_snitch_api_call_error",
+			Help:        "Counter of the number of errors in calls to the DMS API",
+			ConstLabels: prometheus.Labels{"name": operatorName},
+		}),
+		snitchCallDuration: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Name:        "dms_operator_snitch_api_call_duration_seconds",
+			Help:        "Distribution of the timings of API calls to DMS in seconds",
+			ConstLabels: prometheus.Labels{"name": operatorName},
+		}, []string{snitchMethodLabel}),
 	}
+}
+
+// AddAPICall observes metrics for a call to an external API
+// - param controller: The name of the controller making the API call
+// - param req: The HTTP Request structure
+// - param resp: The HTTP Response structure
+// - param duration: The number of seconds the call took.
+func (m *MetricsCollector) AddAPICall(controller string, req *http.Request, resp *http.Response, duration float64) {
+	m.apiCallDuration.With(prometheus.Labels{
+		"controller": controller,
+		"method":     req.Method,
+		"resource":   resourceFrom(req.URL),
+		"status":     resp.Status,
+	}).Observe(duration)
+}
+
+// ObserveReconcile records the duration of the reconcile loop of the operator
+func (m *MetricsCollector) ObserveReconcile(seconds float64) {
+	m.ReconcileDuration.Observe(seconds)
+}
+
+// AddCollector adds a collector to existing metrics
+func (m *MetricsCollector) AddCollector(collector prometheus.Collector) {
+	m.collectors = append(m.collectors, collector)
+}
+
+// RecordSnitchCallDuration records the time taken to make a call to the Dead Man Snitch API
+func (m *MetricsCollector) RecordSnitchCallDuration(duration time.Duration, operation string) {
```

Comment:
nit: It would be nice to make these more consistent.
- Right now the names of the methods are `Add`, `Observe`, and `Record`. Can we pick one?
- We've got a `duration float64`, a `seconds float64`, and a `duration time.Duration` which does `.Seconds()` within the func. Can we pick one?

Reply:
Addressed in #69.